batcher/aligned-batcher/src/lib.rs

```diff
@@ -455,206 +341,177 @@ impl Batcher {
+                addr,
+            )
+            .await;
+
+            return Ok(());
         }
 
-        batch_state.increment_user_proof_count(addr);
-        true
-    }
+        let msg_max_fee = nonced_verification_data.max_fee;
+        let Some(user_min_fee) = batch_state_lock.get_user_min_fee(&addr).await else {
+            std::mem::drop(batch_state_lock);
+            send_message(ws_conn_sink.clone(), ValidityResponseMessage::InvalidNonce).await;
+            return Ok(());
+        };
 
-    /// Handles a message with an expected nonce.
-    /// If the max_fee is valid, it is added to the batch.
-    /// If the max_fee is invalid, a message is sent to the client.
-    /// Returns true if the message was added to the batch, false otherwise.
-    async fn handle_expected_nonce_message(
-        &self,
-        mut batch_state: tokio::sync::MutexGuard<'_, BatchState>,
-        min_fee: U256,
-        nonced_verification_data: NoncedVerificationData,
-        ws_conn_sink: Arc<RwLock<SplitSink<WebSocketStream<TcpStream>, Message>>>,
-        signature: Signature,
-        addr: Address,
-    ) -> bool {
-        let max_fee = nonced_verification_data.max_fee;
-        if max_fee > min_fee {
-            warn!(
-                "Invalid max fee for address {addr}, had fee {:?} < {:?}",
-                min_fee, max_fee
-            );
+        if msg_max_fee > user_min_fee {
+            std::mem::drop(batch_state_lock);
+            warn!("Invalid max fee for address {addr}, had fee {user_min_fee:?} < {msg_max_fee:?}");
             send_message(ws_conn_sink.clone(), ValidityResponseMessage::InvalidMaxFee).await;
-            return false;
+            return Ok(());
         }
 
-        let nonce = nonced_verification_data.nonce;
+        // * ---------------------------------------------------------------------*
+        // *        Add message data into the queue and update user state         *
+        // * ---------------------------------------------------------------------*
 
-        batch_state.user_nonces.insert(addr, nonce + U256::one());
-        batch_state.user_min_fee.insert(addr, max_fee);
+        if let Err(e) = self
+            .add_to_batch(
+                batch_state_lock,
+                nonced_verification_data,
+                ws_conn_sink.clone(),
+                client_msg.signature,
+                addr,
+            )
+            .await
+        {
+            error!("Error while adding entry to batch: {e:?}");
+            send_message(ws_conn_sink, ValidityResponseMessage::AddToBatchError).await;
+            return Ok(());
+        };
 
-        self.add_to_batch(
-            batch_state,
-            nonced_verification_data,
-            ws_conn_sink.clone(),
-            signature,
-            addr,
-        )
-        .await;
+        info!("Verification data message handled");
+        send_message(ws_conn_sink, ValidityResponseMessage::Valid).await;
```

Comment:
We should probably drop early here too.

Reply:
here the lock is passed to the `add_to_batch` function so that function takes ownership of it